[FEATURE] pnpm 11 support

### Problem Statement

I'd like to be able to use pnpm v11 with Mastra. `mastra build` fails on pnpm v11 with `ERR_PNPM_IGNORED_BUILDS` when the project depends on a package with postinstall scripts (e.g. `better-sqlite3`). This happens because the deployer writes incorrect pnpm metadata to the build output, then runs `pnpm install --ignore-workspace` in that directory, so the user's own build policy is dropped.

Related parts:
- `packages/deployer/src/services/deps.ts:148` runs `pnpm install --ignore-workspace` in that output dir, ignoring the user's configuration.
- `packages/deployer/src/bundler/index.ts:85-87` writes `pnpm: { neverBuiltDependencies: [] }` to `.mastra/output/package.json`, which is not respected in v11. This is also a security problem for v10, because `neverBuiltDependencies: []` bypasses the pnpm build allow/deny list. If a user originally disallowed the build script for a dependency for security reasons, then they would silently get hit by the problem since Mastra ignores this and allows all scripts.

It had security problems with v10, but worked without errors since it allowed all. In pnpm v10, `neverBuiltDependencies` and friends got merged into `allowBuilds`, so it doesn't work anymore.

### Proposed Solution

Replace `--ignore-workspace` + `pnpm: { neverBuiltDependencies: [] }` with a self-contained `pnpm-workspace.yaml` written into `.mastra/output/` before install:

- Even an empty `pnpm-workspace.yaml` file makes pnpm pick up that file while searching up the file tree. So, it will replace `--ignore-workspace`. `packages:` specified in the actual project will be ignored, just as the existing code intended.
- Carry over other relevant settings, e.g. `allowBuilds`(v11) and `onlyBuiltDependencies` / `neverBuiltDependencies` / `ignoredBuiltDependencies` (v10) into `pnpm-workspace.yaml`
  - A combination of `pnpm config list --json` (resolves `pnpm-workspace.yaml`, `.npmrc`, and global config) and the parent's `package.json > pnpm` field (which I believe `pnpm config` command does not take into account) should work to derive these
- Stop writing `pnpm: { neverBuiltDependencies: [] }` to the output `package.json` to avoid security problems on v10 and redundant configuration(which might decide to print warnings or errors later) on v11+

### Component

Deployment

### Alternatives Considered

- stay stuck on pnpm v10
- pass `pnpm_config_dangerously_allow_all_builds=true` env variable to make it work, same as Mastra's current security-problematic behavior with pnpm v10
- apply a local patch `@mastra/deployer` with `pnpm patch`

### Example Use Case

Self-explanatory

### Additional Context

_No response_

### Verification

- [x] I have searched the existing issues to make sure this is not a duplicate
- [x] I have provided sufficient context for the team to understand the request

Still reproducing on the latest published mastra. The deployer's pnpm-11 `allowBuilds` block is a **hardcoded 7-package list** that ignores the project's own build-script policy, so any externalized dependency with an install script that isn't one of those 7 fails the deploy install under pnpm 11.

### Environment
- `mastra@1.13.0`, `@mastra/deployer@1.42.0`, `@mastra/core@1.42.0`
- `pnpm@11.7.0`, `node v24.13.0`

### Minimal reproduction
```bash
mkdir repro && cd repro
cat > package.json <<'JSON'
{ "name": "repro", "type": "module",
  "scripts": { "build": "mastra build" },
  "packageManager": "pnpm@11.7.0" }
JSON
cat > pnpm-workspace.yaml <<'YAML'
allowBuilds:
  esbuild: true
  protobufjs: false
  '@google/genai': false
YAML
mkdir -p src/mastra
cat > src/mastra/index.ts <<'TS'
import { Mastra } from "@mastra/core";
import { GoogleGenAI } from "@google/genai";   // make it a real runtime dep
export const genai = new GoogleGenAI({ apiKey: process.env.GOOGLE_API_KEY ?? "x" });
export const mastra = new Mastra({});
TS
pnpm add mastra @mastra/core zod @google/genai
CI=true pnpm run build
```

Result:
```
INFO  Installing dependencies
[ERR_PNPM_IGNORED_BUILDS] Ignored build scripts: @google/genai@2.8.0
ERROR Failed during bundler bundle stage: Process exited with code 1
```

(A vanilla app with no extra native-build deps builds fine — its only build-script dep is `esbuild`, which *is* on the list. The failure appears as soon as you depend on something with an install script that the deployer didn't hardcode; in our real project it's `@google/genai` and `msgpackr-extract`.)

### Root cause
The deployer writes the output `.mastra/output/pnpm-workspace.yaml` from a hardcoded string (in `@mastra/deployer`, both `Bundler.writePackageJson` and `Deps.writePnpmConfig`):
```yaml
allowBuilds:
  bcrypt: true
  esbuild: true
  sharp: true
  protobufjs: true
  workerd: true
  bufferutil: true
  utf-8-validate: true
```
It does not consult the project's own `allowBuilds` / `onlyBuiltDependencies` policy. When an externalized dependency outside this list carries an install script, pnpm 11 — in CI / non-interactive installs — treats the ignored build as a hard error (`ERR_PNPM_IGNORED_BUILDS`), whereas pnpm 10 only warned. pnpm then auto-appends the offender to the file as `'@google/genai': set this to true or false` — not a valid boolean — so re-running never self-heals.

There's no clean userland workaround: the output workspace is regenerated on every build, the failing install runs *inside* `mastra build`, and the project's root `allowBuilds` (or `ignored-built-dependencies[]` in a copied-through `.npmrc`) does not override the generated file — all tested.

### Suggested fix
The deployer should **honor the source project's declared build policy** (`allowBuilds` / `onlyBuiltDependencies`) when generating the output workspace, rather than substituting a fixed 7-package list. Failing that, for build-script deps it externalizes but won't build, it should emit a valid `false` instead of the unparseable `set this to true or false` placeholder — though that risks silently skipping an install script a deployed app actually needs (the original `better-sqlite3` case), so reading the project's policy is the more correct fix. Note the list is duplicated across two code paths, so a fix needs to cover both.

--- a/.changeset/pnpm-policy-preserve.md
+++ b/.changeset/pnpm-policy-preserve.md
@@ -1,0 +1,5 @@
+---
+'@mastra/deployer': patch
+---
+
+Copy pnpm install policy from the source workspace into generated deployer output so pnpm 11 build-script approvals are preserved.

--- a/packages/deployer/src/bundler/index.ts
+++ b/packages/deployer/src/bundler/index.ts
@@ -85,12 +85,6 @@ export abstract class Bundler extends MastraBundler {
         2,
       ),
     );
-
-    // pnpm v11 requires build policy via pnpm-workspace.yaml in the output directory
-    await writeFile(
-      join(outputDirectory, 'pnpm-workspace.yaml'),
-      "packages:\n  - '.'\nallowBuilds:\n  bcrypt: true\n  esbuild: true\n  sharp: true\n  protobufjs: true\n  workerd: true\n  bufferutil: true\n  utf-8-validate: true\n",
-    );
   }
 
   protected createBundler(inputOptions: InputOptions, outputOptions: Partial<OutputOptions> & { dir: string }) {

--- a/packages/deployer/src/services/deps.test.ts
+++ b/packages/deployer/src/services/deps.test.ts
@@ -1,0 +1,23 @@
+import { describe, expect, it } from 'vitest';
+import { copyPnpmWorkspaceSettings } from './deps';
+
+describe('copyPnpmWorkspaceSettings', () => {
+  it('copies pnpm install policy without copying source workspace packages', () => {
+    const output = copyPnpmWorkspaceSettings(
+      `packages:\n  - packages/*\n\ncatalog:\n  react: ^19.0.0\n\nminimumReleaseAge: 1440\nminimumReleaseAgeExclude:\n  - '@mastra/*'\n\nallowBuilds:\n  onnxruntime-node: false\n  node-pty: true\n\npatchedDependencies:\n  foo@1.0.0: patches/foo.patch\n`,
+    );
+
+    expect(output).toBe(
+      `packages:\n  - '.'\n\nminimumReleaseAge: 1440\n\nminimumReleaseAgeExclude:\n  - '@mastra/*'\n\nallowBuilds:\n  onnxruntime-node: false\n  node-pty: true\n`,
+    );
+  });
+
+  it('uses requested architecture over source supportedArchitectures', () => {
+    const output = copyPnpmWorkspaceSettings(
+      `packages:\n  - packages/*\n\nsupportedArchitectures:\n  os: [\"linux\"]\n`,
+      { os: ['darwin'], cpu: ['arm64'] },
+    );
+
+    expect(output).toBe(`packages:\n  - '.'\n\nsupportedArchitectures:\n  os: [\"darwin\"]\n  cpu: [\"arm64\"]\n`);
+  });
+});

--- a/packages/deployer/src/services/deps.ts
+++ b/packages/deployer/src/services/deps.ts
@@ -16,6 +16,69 @@ interface ArchitectureOptions {
   libc?: string[];
 }
 
+const PNPM_CONFIG_KEYS_TO_COPY = new Set([
+  'allowBuilds',
+  'onlyBuiltDependencies',
+  'ignoredBuiltDependencies',
+  'neverBuiltDependencies',
+  'minimumReleaseAge',
+  'minimumReleaseAgeExclude',
+  'trustPolicy',
+  'trustPolicyExclude',
+  'trustPolicyIgnoreAfter',
+  'supportedArchitectures',
+]);
+
+function getTopLevelYamlKey(line: string) {
+  const match = /^(?!\s)([\w-]+):/.exec(line);
+  return match?.[1];
+}
+
+export function copyPnpmWorkspaceSettings(source: string, options: ArchitectureOptions = {}) {
+  const hasArchitecture = Boolean(options.os?.length || options.cpu?.length || options.libc?.length);
+  const lines = source.split(/\r?\n/);
+  const blocks: string[] = [];
+
+  for (let index = 0; index < lines.length; ) {
+    const key = getTopLevelYamlKey(lines[index] ?? '');
+    if (!key) {
+      index += 1;
+      continue;
+    }
+
+    const start = index;
+    index += 1;
+    while (index < lines.length && !getTopLevelYamlKey(lines[index] ?? '')) {
+      index += 1;
+    }
+
+    if (!PNPM_CONFIG_KEYS_TO_COPY.has(key) || (key === 'supportedArchitectures' && hasArchitecture)) {
+      continue;
+    }
+
+    const block = lines.slice(start, index).join('\n').trimEnd();
+    if (block) {
+      blocks.push(block);
+    }
+  }
+
+  if (hasArchitecture) {
+    const architectureBlock = ['supportedArchitectures:'];
+    if (options.os?.length) {
+      architectureBlock.push(`  os: ${JSON.stringify(options.os)}`);
+    }
+    if (options.cpu?.length) {
+      architectureBlock.push(`  cpu: ${JSON.stringify(options.cpu)}`);
+    }
+    if (options.libc?.length) {
+      architectureBlock.push(`  libc: ${JSON.stringify(options.libc)}`);
+    }
+    blocks.push(architectureBlock.join('\n'));
+  }
+
+  return ["packages:\n  - '.'", ...blocks].join('\n\n') + '\n';
+}
+
 export class Deps extends MastraBase {
   private packageManager: PackageManager;
   private rootDir: string;
@@ -90,38 +153,31 @@ export class Deps extends MastraBase {
     });
   }
 
-  private async writePnpmConfig(dir: string, options: ArchitectureOptions) {
+  private findPnpmWorkspaceFile(dir: string): string | null {
     const workspaceYamlPath = path.join(dir, 'pnpm-workspace.yaml');
-
-    const lines: string[] = [
-      'packages:',
-      "  - '.'",
-      'allowBuilds:',
-      '  bcrypt: true',
-      '  esbuild: true',
-      '  sharp: true',
-      '  protobufjs: true',
-      '  workerd: true',
-      '  bufferutil: true',
-      '  utf-8-validate: true',
-      'minimumReleaseAge: 0',
-    ];
-    if (options.os?.length || options.cpu?.length || options.libc?.length) {
-      lines.push('');
-      lines.push('supportedArchitectures:');
-      if (options.os?.length) {
-        lines.push(`  os: ${JSON.stringify(options.os)}`);
-      }
-      if (options.cpu?.length) {
-        lines.push(`  cpu: ${JSON.stringify(options.cpu)}`);
-      }
-      if (options.libc?.length) {
-        lines.push(`  libc: ${JSON.stringify(options.libc)}`);
-      }
+    if (fs.existsSync(workspaceYamlPath)) {
+      return workspaceYamlPath;
     }
-    lines.push('');
 
-    await fsPromises.writeFile(workspaceYamlPath, lines.join('\n'), 'utf-8');
+    const parentDir = path.resolve(dir, '..');
+    if (parentDir !== dir) {
+      return this.findPnpmWorkspaceFile(parentDir);
+    }
+
+    return null;
+  }
+
+  private async writePnpmConfig(dir: string, options: ArchitectureOptions = {}) {
+    const sourceWorkspaceYamlPath = this.findPnpmWorkspaceFile(this.rootDir);
+    const sourceWorkspaceYaml = sourceWorkspaceYamlPath
+      ? await fsPromises.readFile(sourceWorkspaceYamlPath, 'utf-8')
+      : '';
+
+    await fsPromises.writeFile(
+      path.join(dir, 'pnpm-workspace.yaml'),
+      copyPnpmWorkspaceSettings(sourceWorkspaceYaml, options),
+      'utf-8',
+    );
   }
 
   private async writeYarnConfig(dir: string, options: ArchitectureOptions) {
@@ -181,9 +237,7 @@ export class Deps extends MastraBase {
 
     switch (pm) {
       case 'pnpm':
-        if (architecture) {
-          await this.writePnpmConfig(dir, architecture);
-        }
+        await this.writePnpmConfig(dir, architecture);
         break;
       case 'yarn':
         // similar to --ignore-workspace but for yarn

